src/storage/src/render/upsert/types.rs

```diff
@@ -7,85 +7,478 @@
+        Self::Decoded(uv)
+    }
+}
+
+impl StateValue {
+    /// We use a XOR trick in order to accumulate the snapshot without having to store the full
+    /// unconsolidated history in memory. For all (value, diff) updates of a key we track:
+    /// - diff_sum = SUM(diff)
+    /// - checksum_sum = SUM(checksum(bincode(value)) * diff)
+    /// - len_sum = SUM(len(bincode(value)) * diff)
+    /// - value_xor = XOR(bincode(value))
+    ///
+    /// ## Correctness
+    ///
+    /// The method is correct because a well formed upsert snapshot will have for each key:
+    /// - Zero or one updates of the form (cur_value, +1)
+    /// - Zero or more pairs of updates of the form (prev_value, +1), (prev_value, -1)
+    ///
+    /// We are interested in extracting the cur_value of each key and discard all prev_values
+    /// that might be included in the stream. Since the history of prev_values always comes in
+    /// pairs, computing the XOR of those is always going to cancel their effects out. Also,
+    /// since XOR is commutative this property is true independent of the order. The same is
+    /// true for the summations of the length and checksum since the sum will contain the
+    /// unrelated values zero times.
+    ///
+    /// Therefore the accumulators will end up precisely in one of two states:
+    /// 1. diff == 0, checksum == 0, value == [0..] => the key is not present
+    /// 2. diff == 1, checksum == checksum(cur_value) value == cur_value => the key is present
+    ///
+    /// ## Robustness
+    ///
+    /// In the absense of bugs, accumulating the diff and checksum is not required since we know
+    /// that a well formed snapshot always satisfies XOR(bincode(values)) == bincode(cur_value).
+    /// However bugs may happen and so storing 16 more bytes per key to have a very high
+    /// guarantee that we're not decoding garbage is more than worth it.
+    /// The main key->value used to store previous values.
+    #[allow(clippy::as_conversions)]
+    pub fn merge_update(
+        &mut self,
+        value: UpsertValue,
+        diff: mz_repr::Diff,
+        bincode_opts: BincodeOpts,
+        bincode_buffer: &mut Vec<u8>,
+    ) -> bool {
+        if let Self::Snapshotting(Snapshotting {
+            value_xor,
+            len_sum,
+            checksum_sum,
+            diff_sum,
+        }) = self
+        {
+            bincode_buffer.clear();
+            bincode_opts
+                .serialize_into(&mut *bincode_buffer, &value)
+                .unwrap();
+            let len = i64::try_from(bincode_buffer.len()).unwrap();
+
+            *diff_sum += diff;
+            *len_sum += len.wrapping_mul(diff);
+            // Truncation is fine (using `as`) as this is just a checksum
+            *checksum_sum += (seahash::hash(&*bincode_buffer) as i64).wrapping_mul(diff);
+
+            // XOR of even diffs cancel out, so we only do it if diff is odd
+            if diff.abs() % 2 == 1 {
```

Comment:
dumb q: Will `diff` for the single value ever be **not** `+1` or `-1`?

Reply:
Actually should this be comparing `diff_sum` instead of `diff`?

Reply:
Not in the source pipeline!